packages/data-point/src/ReducerPath.js

```diff
@@ -1,36 +1,94 @@
-const { ReducerNative } = require("./ReducerNative");
+const { Reducer } = require("./Reducer");
 
 // Credit:
+// path parsing based off of the work on:
 // https://gomakethings.com/how-to-get-the-value-of-an-object-from-a-specific-path-with-vanilla-js/
 // chris@gomakethings.com
 
-const splitToArrayRegEx = /\[([^}]+)\]/g;
+/**
+ * @type {RegExp} matches simple accessor separators `[`, `]` and `.`
+ */
+const splitToArrayRegEx = /[[,.,\]]/g;
 
-function stringToPath(path) {
+/**
+ * @param {string} pathProperty
+ * @returns {number|string} parsed value
+ */
+function parsePathProperty(pathProperty) {
+  const numberValue = Number(pathProperty);
+  return Number.isNaN(numberValue) ? pathProperty : numberValue;
+}
+
+/**
+ * @param {string} pathSource
+ * @returns {Array<string|number>}
+ */
+function stringToPath(pathSource) {
   const output = [];
 
-  // Split to an array with dot notation
-  path.split(".").forEach(item => {
-    // Split to an array with bracket notation
-    item.split(splitToArrayRegEx).forEach(key => {
-      // Push to the new array
-      if (key.length > 0) {
-        output.push(key);
-      }
-    });
-  });
+  const accessors = pathSource.split(splitToArrayRegEx);
+
+  const accessorsLength = accessors.length;
+
+  // using for loop for faster iteration
+  for (let index = 0; index < accessorsLength; index += 1) {
+    const accessor = accessors[index];
+    if (accessor !== "") {
+      output.push(parsePathProperty(accessor));
+    }
+  }
 
   return output;
 }
 
-function getPath(value, path) {
+/**
+ * @param {string} pathSource
+ * @returns {Object} compiled source
+ */
+function parsePath(pathSource) {
+  const compliedSource = {
+    /**
+     * @type {Array<string|number>} parsed object path
+     */
+    path: [],
+    /**
+     * @type {Boolean} true if crawling should look one level up
+     */
+    moveUp: false
+  };
+
+  let sourceToCompile = pathSource.substr(1);
+
+  if (pathSource.startsWith("$..")) {
+    compliedSource.moveUp = true;
+    sourceToCompile = sourceToCompile.substr(2);
+  }
+
+  if (pathSource === "$.") {
+    compliedSource.path = [];
+  } else {
+    compliedSource.path = stringToPath(sourceToCompile);
+  }
+
+  return compliedSource;
+  // return pathSource === "$." ? [] : stringToPath(pathSource.substr(1));
+}
+
+/**
+ * @param {any} value value to crawl
+ * @param {Array<string|number>} path object path to crawl
+ * @returns {any} resolved path
+ */
+function crawlPath(value, path) {
   let current = value;
 
   // For each item in the path, dig into the object
   const length = path.length;
+
+  // using for loop for faster iteration
   for (let i = 0; i < length; i += 1) {
     // If the item isn't found, return undefined
-    if (!current[path[i]]) return undefined;
+    if (current[path[i]] === undefined) return undefined;
```

Comment:
what if it's null or some other falsy value? wouldn't that cause an error?

Reply:
you are right, fixed. thanks!